README.md
# NAME

Getopt::EX::termcolor - Getopt::EX termcolor module

# SYNOPSIS

    use Getopt::EX::Loader;
    my $rcloader = Getopt::EX::Loader->new(
        BASECLASS => [ 'App::command', 'Getopt::EX' ],
        );

    or

    use Getopt::EX::Long qw(:DEFAULT ExConfigure);
    ExConfigure BASECLASS => [ "App::command", "Getopt::EX" ];

    then

    $ command -Mtermcolor::bg=

# VERSION

Version 1.07

# DESCRIPTION

This is a common module for command using [Getopt::EX](https://metacpan.org/pod/Getopt::EX) to manipulate
system dependent terminal color.

Actual action is done by sub-module under [Getopt::EX::termcolor](https://metacpan.org/pod/Getopt::EX::termcolor),
such as [Getopt::EX::termcolor::Apple\_Terminal](https://metacpan.org/pod/Getopt::EX::termcolor::Apple_Terminal).

Each sub-module is expected to have `&get_color` function which
return the list of RGB values for requested name, but currently name
`background` is only supported.  Each RGB values are expected in a
range of 0 to 255 by default.  If the first entry of the list is a
HASH reference, it may include maximum number indication like `{
max => 65535 }`.

Terminal luminance is calculated from RGB values by this equation and
produces decimal value from 0 to 100.

    ( 30 * R + 59 * G + 11 * B ) / MAX

If the environment variable `TERM_BGCOLOR` is defined, it is used as
a background RGB value without calling sub-modules.  RGB value is a
combination of integer described in 24bit/12bit hex, 24bit decimal or
6x6x6 216 color format.  RGB color notation is compatible with
[Getopt::EX::Colormap](https://metacpan.org/pod/Getopt::EX::Colormap).

    24bit hex     #000000 .. #FFFFFF
    12bit hex     #000    .. #FFF
    24bit decimal 0,0,0   .. 255,255,255
    6x6x6 216     000     .. 555

You can set `TERM_BGCOLOR` in you start up file of shell.  This
module has utility function `bgcolor` which can be used like this:

    export TERM_BGCOLOR=`perl -MGetopt::EX::termcolor=bgcolor -e bgcolor`
    : ${TERM_BGCOLOR:=#FFFFFF}

# MODULE FUNCTION

- **bg**

    Call this function with module option:

        $ command -Mtermcolor::bg=

    If the terminal luminance is unknown, nothing happens.  Otherwise, the
    module insert **--light-terminal** or **--dark-terminal** option
    according to the luminance value.

    You can change the behavior by optional parameters:

        threshold : threshold of light/dark  (default 50)
        default   : default luminance value  (default none)
        light     : light terminal option    (default "--light-terminal")
        dark      : dark terminal option     (default "--dark-terminal")

    Use like this:

        option default \
            -Mtermcolor::bg(default=100,light=--light,dark=--dark)

# SEE ALSO

[Getopt::EX](https://metacpan.org/pod/Getopt::EX)

[Getopt::EX::termcolor::Apple\_Terminal](https://metacpan.org/pod/Getopt::EX::termcolor::Apple_Terminal)

[Getopt::EX::termcolor::iTerm](https://metacpan.org/pod/Getopt::EX::termcolor::iTerm)

[Getopt::EX::termcolor::XTerm](https://metacpan.org/pod/Getopt::EX::termcolor::XTerm)

# AUTHOR

Kazumasa Utashiro

# LICENSE

Copyright 2020-2021 Kazumasa Utashiro.

You can redistribute it and/or modify it under the same terms
as Perl itself.


Changes
Revision history for Getopt-EX extension Getopt-EX-termcolor

{{$NEXT}}

1.08 2022-03-18T02:59:05Z

    - Improve XTerm terminal interface.
      Try:
        perl `perldoc -l Getopt::EX::termcolor::XTerm`

1.07 2020-07-06T01:45:00Z

    - Eliminate indirect object call.

1.06 2020-05-14T07:04:58Z

    - Allow empty option setting for light/dark.
    - Make XTerm timeout longer.

1.05 2020-04-28T00:45:35Z

    - Accept anything start with xterm as XTerm.

1.04 2020-04-23T15:45:42Z

    - Fix bug about $ENV{TERM}.
    - Stop to depend on App::cdif::Command.

1.03 2020-04-23T08:13:28Z

    - Implement XTerm.pm using operating system command sequence.
    - Introduce environment TERM_BGCOLOR and recommend to use.
    - RGB default maximum value is now 255x255x255.

1.02 2020-04-18T02:49:50Z

    - Change the interface to call &bg with module option.
    - Use TERM_BRIGHTNESS rather than BRIGHTNESS.

1.01 2020-04-16T06:15:00Z

    - Rename Getopt::EX::autocolor to Getopt::EX::termcolor.

      App::sdif already have -Mautocolor module.  When upgrading them
      to use Getopt::EX module, there should be conflict with previously
      installed autocolor.pm under the command library.  So I decided to
      change the name.

0.04 2020-04-15T10:43:41Z

    - Fix bug in rgb_to_brightness().
    - Make iTerm.pm to work.

0.03 2020-04-15T09:32:32Z

    - Remove wrong dependency to App::Greple.

0.02 2020-04-15T09:14:13Z

    - Improve customizability.
    - Require Getopt::EX v1.16.0 to use &finalize function.

0.01 2020-04-13T07:57:22Z

    - original version



lib/Getopt/EX/autocolor.pm
=encoding utf-8

=head1 NAME

Getopt::EX::autocolor - Getopt::EX autocolor module

=head1 SYNOPSIS

    use Getopt::EX::Loader;
    my $rcloader = new Getopt::EX::Loader
        BASECLASS => [ 'App::command', 'Getopt::EX' ];

    $ command -Mautocolor

=head1 VERSION

Version 0.03

=head1 DESCRIPTION

This is a common module for command using L<Getopt::EX> to set system
dependent autocolor option.

Actual action is done by sub-module under L<Getopt::EX::autocolor>,
such as L<Getopt::EX::autocolor::Apple_Terminal>.

Each sub-module is expected to have C<&brightness> function which
returns integer value between 0 and 100.  If the sub-module was found
and C<&brightness> function exists, its result is taken as a
brightness of the terminal.

However, if the environment variable C<BRIGHTNESS> is defined, its
value is used as a brightness without calling sub-modules.  The value
of C<BRIGHTNESS> is expected in range of 0 to 100.

If the brightness can not be taken, nothing happens.  Otherwise, the
module insert B<--light-terminal> or B<--dark-terminal> option
according to the brightness value.  These options are defined as
C$<move(0,0)> in this module and do nothing.  They can be overridden
by other module or user definition.

You can change the behavior of this module by calling C<&set> function
with module option.  It takes some parameters and they override
default values.

    threshold : threshold of light/dark  (default 50)
    default   : default brightness value (default none)
    light     : light terminal option    (default "--light-terminal")
    dark      : dark terminal option     (default "--dark-terminal")

For example, use like:

    option default \
        -Mautocolor::set(default=100,light=--light,dark=--dark) \

=head1 FUNCTIONS

=over 7

=item B<rgb_to_brightness>

This exportable function caliculates brightness (luminane) from RGB
values.  It accepts three parameters of 0 to 65535 integer.

Maximum value can be specified by the first hash argument.

    rgb_to_brightness( { max => 255 }, 255, 255, 255);

Brightness is caliculated from RGB values by next equation.

    Y = 0.30 * R + 0.59 * G + 0.11 * B

=back

=head1 SEE ALSO

L<Getopt::EX>

L<Getopt::EX::autocolor::Apple_Terminal>

L<Getopt::EX::autocolor::iTerm>

=head1 AUTHOR

Kazumasa Utashiro

=head1 LICENSE

Copyright (C) 2020 Kazumasa Utashiro.

You can redistribute it and/or modify it under the same terms
as Perl itself.

=cut

package Getopt::EX::autocolor;

use v5.14;
use strict;
use warnings;
use Data::Dumper;

our $VERSION = "0.03";

use Exporter 'import';
our @EXPORT      = qw();
our %EXPORT_TAGS = ();
our @EXPORT_OK   = qw(rgb_to_brightness);

our %param = (
    light => "--light-terminal",
    dark  => "--dark-terminal",
    default => undef,
    threshold => 50,
    );

sub rgb_to_brightness {
    my $opt = ref $_[0] ? shift : {};
    my $max = $opt->{max} || 65535;
    my($r, $g, $b) = @_;
    int(($r * 30 + $g * 59 + $b * 11) / $max); # 0 .. 100
}

sub call(&@) { $_[0]->(@_[1..$#_]) }

sub finalize {
    my $mod = shift;

    # default to do nothing.
    $mod->setopt($param{light} => '$<move(0,0)>');
    $mod->setopt($param{dark}  => '$<move(0,0)>');

    my $brightness = call {
	my $v = $ENV{BRIGHTNESS};
	if (defined $v && $v =~ /^\d+$/
	    && 0 <= $v && $v <= 100
	    ) {
	    return $v;
	}
	if (my $term_program = $ENV{TERM_PROGRAM}) {
	    my $submod = $term_program =~ s/\.app$//r;
	    my $mod = __PACKAGE__ . "::$submod";
	    my $brightness = "$mod\::brightness";
	    no strict 'refs';
	    if (eval "require $mod" and defined &$brightness) {
		my $v = &$brightness;
		if (0 <= $v and $v <= 100) {
		    return $v;
		}
	    }
	}
	undef;
    };

    $brightness //= $param{default} // return;

    $mod->setopt(default =>
		 $brightness > $param{threshold}
		 ? $param{light}
		 : $param{dark});
}

use List::Util qw(pairgrep);

sub set {
    my %arg = @_;
    %param = (%param, pairgrep { exists $param{$a} } %arg);
}

1;

__DATA__

#  LocalWords:  autocolor


lib/Getopt/EX/termcolor/XTerm.pm
=head1 NAME

Getopt::EX::termcolor::XTerm

=head1 SYNOPSIS

use Getopt::EX::termcolor::XTerm;

=head1 DESCRIPTION

This is a L<Getopt::EX::termcolor> module for XTerm.

=head1 SEE ALSO

L<Getopt::EX::termcolor>

L<https://www.xfree86.org/current/ctlseqs.html>

=cut

package Getopt::EX::termcolor::XTerm;

use v5.14;
use strict;
use warnings;

use Exporter 'import';
our @EXPORT_OK = qw(test);

use Carp;
use Data::Dumper;
use IO::Handle;
use Term::ReadKey;

use Getopt::EX::termcolor;

sub get_colors {
    map {
	my @rgb = get_color($_);
	@rgb ? undef : [ @rgb ];
    } @_;
}

my %alias = qw(
    foreground text_foreground
    background text_background
    );

sub get_color {
    my $res = lc shift;
    $res = $alias{$res} if $alias{$res};
    return color_rgb($res);
}

our $debug = $ENV{DEBUG_GETOPTEX};

use constant {
    CSI => "\e[", # Control Sequence Introducer
    OSC => "\e]", # Operating System Command
};

sub osc_command {
    my($Ps, $Pt) = @_;
    OSC . "$Ps;$Pt" . "\a";
}

use List::Util qw(pairmap);

my %oscPs = pairmap { $b => $a, lc $b => $a } qw(
    10 text_foreground
    11 text_background
    12 text_cursor
    13 mouse_foreground
    14 mouse_background
    15 Tektronix_foreground
    16 Tektronix_background
    17 highlight_background
    18 Tektronix_cursor
    19 highlight_foreground
    );

sub uncntrl {
    $_[0] =~ s/([^\040-\176])/sprintf "\\%03o", ord $1/gaer;
}

# OSC Set Text Parameter
sub osc_stp {
    my $name = shift;
    my $color = @_ ? shift : '?';
    my $Ps = $oscPs{$name} or croak;
    osc_command $Ps, $color;
}

my $osc_st_re = qr/[\a\x9c]|\e\\/;
my $osc_answer_re = qr/\e\]\d+;(?<answer>[\x08-\x13\x20-\x7d]*)(?:$osc_st_re)/;

sub osc_answer {
    $_[0] =~ $osc_answer_re and $+{answer};
}

sub ask {
    my $request = shift;
    if ($debug) {
	printf STDERR "[%s] Request: %s\n",
	    __PACKAGE__,
	    uncntrl $request;
    }
    open my $tty, "+<", "/dev/tty" or return;
    ReadMode "cbreak", $tty;
    printflush $tty $request;
    my $timeout = 0.01;
    my $answer = '';
    while (defined (my $key = ReadKey $timeout, $tty)) {
	if (0 and $debug) {
	    printf STDERR "[%s] ReadKey: \"%s\"\n",
		__PACKAGE__,
		$key =~ /\P{Cc}/ ? $key : uncntrl $key;
	}
	$answer .= $key;
	last if $answer =~ /$osc_st_re\z/;
    }
    if ($debug) {
	printf STDERR "[%s] Answer:  %s\n",
	    __PACKAGE__,
	    uncntrl $answer;
    }
    ReadMode "restore", $tty;
    return $answer;
}

use List::Util qw(max);

sub color_rgb {
    my $name = shift;
    my $rgb = osc_answer ask osc_stp $name;
    my @rgb = $rgb =~ m{rgb:([\da-f]+)/([\da-f]+)/([\da-f]+)}i or return;
    my $max = (2 ** (length($1) * 4)) - 1;
    my @opt = $max == 255 ? () : ( { max => $max } );
    ( @opt, map { hex } @rgb );
}

do { test() } if __FILE__ eq $0;

sub test {
    my @name = qw(
	  text_foreground
	  text_background
	  text_cursor
	  mouse_foreground
	  mouse_background
	  Tektronix_foreground
	  Tektronix_background
	  highlight_background
	  Tektronix_cursor
	  highlight_foreground
	);
    local $Data::Dumper::Indent = 1;
    local $Data::Dumper::Terse = 1;
    my $max = max map { length } @name;
    for my $name (@name) {
	my @rgb = color_rgb($name);
	printf "%*s: %s",
	    $max, $name,
	    @rgb ? Dumper(\@rgb)=~s/\n(?!\z)\s*/ /gr : "n/a\n";
    }
}

1;


lib/Getopt/EX/termcolor/iTerm.pm
=head1 NAME

Getopt::EX::termcolor::iTerm

=head1 SYNOPSIS

command -Mtermcolor::iTerm

=head1 DESCRIPTION

This is a L<Getopt::EX::termcolor> module for iTerm.

=head1 SEE ALSO

L<Getopt::EX::termcolor>

=cut

package Getopt::EX::termcolor::iTerm;

use strict;
use warnings;
use Data::Dumper;

use Getopt::EX::termcolor qw(rgb_to_luminance);
use App::cdif::Command::OSAscript;

{
    no warnings 'once';
    use Getopt::EX::Colormap;
    $Getopt::EX::Colormap::RGB24 = 1;
}

our $debug = 0;

sub get_colors {
    map {
	my @rgb = get_color($_);
	@rgb ? undef : [ @rgb ];
    } @_;
}

sub get_color {
    my $cat = lc shift;
    if ($cat eq 'background') {
	return background_rgb();
    }
    ();
}

my %script = (

'background' => <<END,
tell application "iTerm"
	tell current session of current window
		background color
	end tell
end tell
END

    );

sub background_rgb {
    my $script = $script{background};
    my $result = App::cdif::Command::OSAscript->new->exec($script);
    my @rgb = $result =~ /(\d+)/g;
    @rgb == 3 ? ( { max => 65535 }, @rgb) : ();
}

1;

__DATA__


lib/Getopt/EX/termcolor.pm
=encoding utf-8

=head1 NAME

Getopt::EX::termcolor - Getopt::EX termcolor module

=head1 SYNOPSIS

    use Getopt::EX::Loader;
    my $rcloader = Getopt::EX::Loader->new(
        BASECLASS => [ 'App::command', 'Getopt::EX' ],
        );

    or

    use Getopt::EX::Long qw(:DEFAULT ExConfigure);
    ExConfigure BASECLASS => [ "App::command", "Getopt::EX" ];

    then

    $ command -Mtermcolor::bg=

=head1 VERSION

Version 1.07

=head1 DESCRIPTION

This is a common module for command using L<Getopt::EX> to manipulate
system dependent terminal color.

Actual action is done by sub-module under L<Getopt::EX::termcolor>,
such as L<Getopt::EX::termcolor::Apple_Terminal>.

Each sub-module is expected to have C<&get_color> function which
return the list of RGB values for requested name, but currently name
C<background> is only supported.  Each RGB values are expected in a
range of 0 to 255 by default.  If the first entry of the list is a
HASH reference, it may include maximum number indication like C<< {
max => 65535 } >>.

Terminal luminance is calculated from RGB values by this equation and
produces decimal value from 0 to 100.

    ( 30 * R + 59 * G + 11 * B ) / MAX

=begin comment

If the environment variable C<TERM_LUMINANCE> is defined, its value is
used as a luminance without calling sub-modules.  The value of
C<TERM_LUMINANCE> is expected in range of 0 to 100.

=end comment

If the environment variable C<TERM_BGCOLOR> is defined, it is used as
a background RGB value without calling sub-modules.  RGB value is a
combination of integer described in 24bit/12bit hex, 24bit decimal or
6x6x6 216 color format.  RGB color notation is compatible with
L<Getopt::EX::Colormap>.

    24bit hex     #000000 .. #FFFFFF
    12bit hex     #000    .. #FFF
    24bit decimal 0,0,0   .. 255,255,255
    6x6x6 216     000     .. 555

You can set C<TERM_BGCOLOR> in you start up file of shell.  This
module has utility function C<bgcolor> which can be used like this:

    export TERM_BGCOLOR=`perl -MGetopt::EX::termcolor=bgcolor -e bgcolor`
    : ${TERM_BGCOLOR:=#FFFFFF}

=head1 MODULE FUNCTION

=over 7

=item B<bg>

Call this function with module option:

    $ command -Mtermcolor::bg=

If the terminal luminance is unknown, nothing happens.  Otherwise, the
module insert B<--light-terminal> or B<--dark-terminal> option
according to the luminance value.

You can change the behavior by optional parameters:

    threshold : threshold of light/dark  (default 50)
    default   : default luminance value  (default none)
    light     : light terminal option    (default "--light-terminal")
    dark      : dark terminal option     (default "--dark-terminal")

Use like this:

    option default \
        -Mtermcolor::bg(default=100,light=--light,dark=--dark)

=back

=head1 SEE ALSO

L<Getopt::EX>

L<Getopt::EX::termcolor::Apple_Terminal>

L<Getopt::EX::termcolor::iTerm>

L<Getopt::EX::termcolor::XTerm>

=head1 AUTHOR

Kazumasa Utashiro

=head1 LICENSE

Copyright 2020-2021 Kazumasa Utashiro.

You can redistribute it and/or modify it under the same terms
as Perl itself.

=cut

use v5.14;
package Getopt::EX::termcolor;

our $VERSION = '1.07';

use warnings;
use Carp;
use Data::Dumper;

use Exporter 'import';
our @EXPORT      = qw();
our @EXPORT_OK   = qw(rgb_to_luminance rgb_to_brightness luminance bgcolor);
our %EXPORT_TAGS = (all => [ @EXPORT_OK ]);

#
# For backward compatibility.
#
sub rgb_to_brightness {
    goto &rgb_to_luminance;
}

sub rgb_to_luminance {
    @_ or return;
    my $opt = ref $_[0] ? shift : {};
    my $max = $opt->{max} || 255;
    my($r, $g, $b) = @_;
    use integer;
    ($r * 30 + $g * 59 + $b * 11) / $max; # 0 .. 100
}

my $mod;
my $argv;

sub initialize {
    ($mod, $argv) = @_;
    set_luminance();
}

our $debug = 0;

sub debug {
    $debug ^= 1;
}

sub call_mod_sub {
    my($mod, $name, @arg) = @_;
    my $call = "$mod\::$name";
    if (eval "require $mod" and defined &$call) {
	no strict 'refs';
	$call->(@arg);
    } else {
	if ($@ !~ /^Can't locate /) {
	    croak $@;
	}
    }
}

sub rgb255 {
    use integer;
    my $opt = ref $_[0] ? shift : {};
    my $max = $opt->{max} // 255;
    map { $_ * 255 / $max } @_;
}

sub get_rgb {
    my $cat = shift;
    my @rgb;
  RGB:
    {
	# TERM=xterm
	if (($ENV{TERM} // '') =~ /^xterm/) {
	    my $mod = __PACKAGE__ . "::XTerm";
	    @rgb = call_mod_sub $mod, 'get_color', $cat;
	    last if @rgb >= 3;
	}
	# TERM_PROGRAM
	if (my $term_program = $ENV{TERM_PROGRAM}) {
	    warn "TERM_PROGRAM=$ENV{TERM_PROGRAM}\n" if $debug;
	    my $submod = $term_program =~ s/\.app$//r;
	    my $mod = __PACKAGE__ . "::$submod";
	    @rgb = call_mod_sub $mod, 'get_color', $cat;
	    last if @rgb >= 3;
	}
	return ();
    }
  GOTCHA:
    rgb255 @rgb;
}

sub set_luminance {
    my $luminance;
    if (defined $ENV{TERM_LUMINANCE}) {
	warn "TERM_LUMINANCE=$ENV{TERM_LUMINANCE}\n" if $debug;
	return;
    }
    if ("BACKWARD COMPATIBILITY") {
	if (defined (my $env = $ENV{BRIGHTNESS})) {
	    warn "BRIGHTNESS=$env\n" if $debug;
	    $ENV{TERM_LUMINANCE} = $env;
	    return;
	}
    }
    if (my $bgcolor = $ENV{TERM_BGCOLOR}) {
	warn "TERM_BGCOLOR=$bgcolor\n" if $debug;
	if (my @rgb = parse_rgb($bgcolor)) {
	    $luminance = rgb_to_luminance @rgb;
	} else {
	    warn "Invalid format: TERM_BGCOLOR=$bgcolor\n";
	}
    } else {
	$luminance = get_luminance();
    }
    $ENV{TERM_LUMINANCE} = $luminance // return;
}

sub get_luminance {
    rgb_to_luminance get_rgb "background";
}

use List::Util qw(pairgrep);

#
# FOR BACKWARD COMPATIBILITY
# DEPELICATED IN THE FUTURE
#
sub set { goto &bg }

my %bg_param = (
    light => "--light-terminal",
    dark  => "--dark-terminal",
    default => undef,
    threshold => 50,
    );

sub bg {
    my %param =
	(%bg_param, pairgrep { exists $bg_param{$a} } @_);
    my $luminance =
	$ENV{TERM_LUMINANCE} // $param{default} // return;
    my $option = $luminance > $param{threshold} ?
	$param{light} : $param{dark}
    or return;

#   $mod->setopt($option => '$<ignore>');
    $mod->setopt(default => $option);
}

sub parse_rgb {
    my $rgb = shift;
    my @rgb = do {
	if    ($rgb =~ /^\#?([\da-f]{2})([\da-f]{2})([\da-f]{2})$/i) {
	    map { hex } $1, $2, $3;
	}
	elsif ($rgb =~ /^\#([\da-f])([\da-f])([\da-f])$/i) {
	    map { 0x11 * hex } $1, $2, $3;
	}
	elsif ($rgb =~ /^([0-5])([0-5])([0-5])$/) {
	    map { 0x33 * int } $1, $2, $3;
	}
	elsif ($rgb =~ /^(\d+),(\d+),(\d+)$/) {
	    map { int } $1, $2, $3;
	}
	else {
	    return ();
	}
    };
    @rgb;
}

sub luminance {
    my $v = get_luminance() // return;
    say $v;
}

sub bgcolor {
    my @rgb = get_rgb "background" or return;
    printf "#%02X%02X%02X\n", @rgb;
}

1;

__DATA__

#  LocalWords:  termcolor RGB
